system_tests/snap_sync_test.go

```diff
@@ -0,0 +1,146 @@
+		Require(t, err)
+		headerNodeB, err := nodeB.Client.HeaderByNumber(ctx, nil)
+		Require(t, err)
+		if header.Number.Cmp(headerNodeB.Number) == 0 {
+			break
+		} else {
+			<-time.After(10 * time.Millisecond)
+		}
+	}
+
+	batchCount, err := builder.L2.ConsensusNode.InboxTracker.GetBatchCount()
+	Require(t, err)
+	delayedCount, err := builder.L2.ConsensusNode.InboxTracker.GetDelayedCount()
+	Require(t, err)
+	// Last batch is batchCount - 1, so prev batch is batchCount - 2
+	prevBatchMetaData, err := builder.L2.ConsensusNode.InboxTracker.GetBatchMetadata(batchCount - 2)
+	Require(t, err)
+	prevMessage, err := builder.L2.ConsensusNode.TxStreamer.GetMessage(prevBatchMetaData.MessageCount - 1)
+	Require(t, err)
+	// Create a config with snap sync enabled and same database directory as the 2nd node
+	nodeConfig := builder.nodeConfig
+	nodeConfig.SnapSyncTest.Enabled = true
+	nodeConfig.SnapSyncTest.BatchCount = batchCount
+	nodeConfig.SnapSyncTest.DelayedCount = delayedCount
+	nodeConfig.SnapSyncTest.PrevDelayedRead = prevMessage.DelayedMessagesRead
+	nodeConfig.SnapSyncTest.PrevBatchMessageCount = uint64(prevBatchMetaData.MessageCount)
+	// Cleanup the message data of 2nd node, but keep the block state data.
+	// This is to simulate a snap sync environment where we’ve just gotten the block state but don’t have any messages.
+	err = os.RemoveAll(nodeB.ConsensusNode.Stack.ResolvePath("arbitrumdata"))
+	Require(t, err)
+
+	// Cleanup the 2nd node to release the database lock
+	cleanupB()
+	// New node with snap sync enabled, and the same database directory as the 2nd node but with no message data.
+	nodeC, cleanupC := builder.Build2ndNode(t, &SecondNodeParams{stackConfig: nodeB.ConsensusNode.Stack.Config(), nodeConfig: nodeConfig})
+	defer cleanupC()
+
+	// Create transactions till batch count is 20
+	for {
+		tx := builder.L2Info.PrepareTx("Faucet", "BackgroundUser", builder.L2Info.TransferGas, big.NewInt(1), nil)
+		err := builder.L2.Client.SendTransaction(ctx, tx)
+		Require(t, err)
+		_, err = builder.L2.EnsureTxSucceeded(tx)
+		Require(t, err)
+		count, err := builder.L2.ConsensusNode.InboxTracker.GetBatchCount()
+		Require(t, err)
+		if count > 20 {
+			break
+		}
+	}
+	// Wait for nodeB to sync up to the first node
+	for {
+		count, err := builder.L2.ConsensusNode.InboxTracker.GetBatchCount()
+		Require(t, err)
+		countNodeC, err := nodeC.ConsensusNode.InboxTracker.GetBatchCount()
+		Require(t, err)
+		if count == countNodeC {
+			// Once the node is synced up, check if the batch metadata is the same for the last batch
+			// This is to ensure that the snap sync worked correctly
+			metadata, err := builder.L2.ConsensusNode.InboxTracker.GetBatchMetadata(count - 1)
+			Require(t, err)
+			metadataNodeC, err := nodeC.ConsensusNode.InboxTracker.GetBatchMetadata(countNodeC - 1)
+			Require(t, err)
+			if metadata != metadataNodeC {
+				t.Error("Batch metadata mismatch")
+			}
+			break
+		} else {
+			<-time.After(10 * time.Millisecond)
```

Comment:
add
```continue```
here - so we'll circle back to checking count and not fall to checking blocks after count didn't pass

Reply:
Thanks for catching this, done!